shared/src/commonMain/kotlin/dev/johnoreilly/confetti/ApolloClientCache.kt

```diff
@@ -19,64 +27,71 @@ interface TokenProvider {
     suspend fun token(forceRefresh: Boolean): String?
 }
 
-/**
- * A factory function which creates an anonymous instance for [TokenProvider] where [getToken] is
- * used when [TokenProvider.token] is invoked.
- */
-inline fun TokenProvider(
-    crossinline getToken: suspend (forceRefresh: Boolean) -> String?,
-): TokenProvider =
-    object : TokenProvider {
-        override suspend fun token(forceRefresh: Boolean): String? = getToken(forceRefresh)
-    }
+
+class TokenProviderContext(val tokenProvider: TokenProvider): ExecutionContext.Element{
+    override val key: ExecutionContext.Key<*>
+        get() = Key
+    companion object Key: ExecutionContext.Key<TokenProviderContext>
+}
 
 class ApolloClientCache : KoinComponent {
     val _clients = mutableMapOf<String, ApolloClient>()
     val mutex = Mutex(false)
-    private val tokenProvider = get<TokenProvider>()
 
-    private val httpInterceptor = object : HttpInterceptor {
-        override suspend fun intercept(
-            request: HttpRequest,
-            chain: HttpInterceptorChain
-        ): HttpResponse {
-            val token = tokenProvider.token(false)
+    private val tokenProviderInterceptor = object : ApolloInterceptor {
+        override fun <D : Operation.Data> intercept(
+            request: ApolloRequest<D>,
+            chain: ApolloInterceptorChain
+        ): Flow<ApolloResponse<D>> {
+            val tokenProvider = request.executionContext[TokenProviderContext]?.tokenProvider
+            if (tokenProvider == null) {
+                return chain.proceed(request)
+            }
+            val token = runBlocking {
+                tokenProvider.token(false)
+            }
             if (token == null) {
                 return chain.proceed(request)
             }
-            val newRequest =
-                request.newBuilder().addHeader("Authorization", "Bearer $token").build()
+            val newRequest = request.newBuilder().addHttpHeader("Authorization", "Bearer $token").build()
             return chain.proceed(newRequest)
         }
-
     }
 
+    suspend fun getClient(conference: String, uid: String?): ApolloClient {
+        return mutex.withLock {
+            _clients.getOrPut("$conference-$uid") {
+                clientFor(conference, uid, true)
+            }
+        }
+    }
     suspend fun getClient(conference: String): ApolloClient {
         return mutex.withLock {
             _clients.getOrPut(conference) {
-                clientFor(conference, conference != "all")
+                clientFor(conference, "none", conference != "all")
             }
         }
     }
 
     private fun clientFor(
         conference: String,
+        uid: String?,
         writeToCacheAsynchronously: Boolean
     ): ApolloClient {
-        val sqlNormalizedCacheFactory = SqlNormalizedCacheFactory(getDatabaseName(conference))
+        val sqlNormalizedCacheFactory = SqlNormalizedCacheFactory(getDatabaseName(conference, uid))
```

Comment:
Is this basically to avoid accidental contamination across users?  Could we do something more extreme, like restart the app (if possible) and delete the DB, to avoid maintaining per user files?